torch/testing/_asserts.py

```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+
+    If both tolerances, :attr:`rtol` and :attr:`rtol`, are ``0``, asserts that :attr:`a` and :attr:`b` are bitwise
+    equal.
 
-        To assert that the values in two tensors are are close but are not required to be bitwise equal, use
-        :func:`assert_tensors_close` instead.
+    Optionally, checks that some attributes of both tensors are equal.
+
+    For a description of the parameters see :func:`assert_equal`.
+
+    Returns:
+        Optional[Exception]: If checks did not pass.
     """
     exc: Optional[Exception] = _check_are_tensors(a, b)
     if exc:
-        raise exc
+        return exc
 
     exc = _check_supported_tensors(a, b)
     if exc:
-        raise exc
+        return exc
+
+    if (rtol is None) ^ (atol is None):
+        # We require both tolerance to be omitted or specified, because specifying only one might lead to surprising
+        # results. Imagine setting atol=0.0 and the tensors still match because rtol>0.0.
+        return UsageError(
+            f"Both 'rtol' and 'atol' must be omitted or specified, " f"but got rtol={rtol} and atol={atol} instead."
+        )
+    elif rtol is None:
+        rtol, atol = _get_default_rtol_and_atol(a, b)
 
     exc = _check_attributes_equal(a, b, check_device=check_device, check_dtype=check_dtype, check_stride=check_stride)
     if exc:
-        raise exc
+        return exc
     a, b = _equalize_attributes(a, b)
 
-    exc = _check_values_equal(a, b)
+    if (rtol == 0.0) and (atol == 0.0):
+        exc = _check_values_equal(a, b)
+    else:
+        exc = _check_values_close(a, b, rtol=rtol, atol=atol)
+    if exc:
+        return exc
+
+    return None
+
+
+def _check_by_type(
+    a: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
+    b: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
+    check_tensors: Callable[[Tensor, Tensor], Optional[Exception]],
+) -> Optional[Exception]:
+    """Delegates tensor checking based on the inputs types.
+
+    Currently supports pairs of
+
+    - :class:`Tensor`'s,
+    - :class:`~collections.abc.Sequence`'s of :class:`Tensor`'s, and
+    - :class:`~collections.abc.Mapping`'s of :class:`Tensor`'s.
+
+    Args:
+        a (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): First input.
+        b (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Second input.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if two tensors match.
+            In case they mismatch should return an :class:`Exception` with an expressive error message.
+
+    Returns:
+        (Optional[Exception]): :class:`UsageError` if the inputs types are unsupported. Additionally, any exception
+        returned by :attr:`check_tensors`.
+    """
+    # _check_are_tensors() returns nothing in case both inputs are tensors and an exception otherwise. Thus, the logic
+    # is inverted here.
+    are_tensors = not _check_are_tensors(a, b)
```

Comment:
Might as well check explicitly for this being None

Is it weird that it doesn't return a tuple (True/False, optional exception)?

Reply:
Returning `ok:bool, exc: Optional[Exception]` has redundant information since the output would always be one of `True, Exception` or `False, None`. Thus, `ok == not(bool(exc))`. This discussion is moot anyway, because #55786 removes this check in order to allow other input types.